### ISSUE TITLE
Fix false positives for inside calc function in to length-zero-no-unit

### ISSUE THREAD

**Issue by Jessidhia** [stylelint/stylelint#3037]
`calc()` does not allow certain operands to be unitless. For example, the operands for `+` and `-` must _always_ have a unit, even if the value is `0`. Conversely, at least one of the operands for `*` must be unitless, and the right hand side operand of `/` must be unitless; but those are unlikely to be flagged by this as you're normally not using `0` there.

The actual CSS is a bit more complicated as it involves custom variables and was working around a bug, but the following reduced test case also reproduces the issue:

```css
.foo {
  margin-left: calc(0px - 10px);
}
```

Compare results on a browser:

<img width="221" alt="image" src="https://user-images.githubusercontent.com/73085/33253487-05c532d2-d387-11e7-8b68-5f0647ec586f.png">
<img width="213" alt="image" src="https://user-images.githubusercontent.com/73085/33253494-0e6b1eba-d387-11e7-86ff-1b2ba0e5e4f2.png">

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
    "extends": "stylelint-config-standard"
}
```

> Which version of stylelint are you using?

e.g. `8.3.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint  'src/**/*.{scss,css}'`

> What did you expect to happen?

No warning as unitless values are not valid as operands to certain `calc` operators.

> What actually happened (e.g. what warnings or errors you are getting)?

```shell
 292:12  ✖  Unexpected unit                               length-zero-no-unit
```

---

As a workaround, I've converted all of the `calc()` that were subtracting from `0px` to multiply their results by `-1` instead.

**Comment by hudochenkov** [comment]
Thanks for the report and for using the template.

`width: calc(0 + 5px)` is invalid, but `width: calc(0 * 5px)` is valid according to [spec](https://drafts.csswg.org/css-values-3/#calc-notation):

> Note: Because <a href="https://drafts.csswg.org/css-syntax-3/#typedef-number-token">&lt;number-token&gt;</a>s are always interpreted as <a href="https://drafts.csswg.org/css-values-3/#number-value">&lt;number&gt;</a>s or <a href="https://drafts.csswg.org/css-values-3/#integer-value">&lt;integer&gt;</a>s, "unitless 0" <a href="https://drafts.csswg.org/css-values-3/#length-value" title="Expands to: em | advance measure | ch | in | cm | vmin | pc | vh | px | rem | pt | q | mm | ex | vw | vmax">&lt;length&gt;</a>s aren’t supported in <a href="https://drafts.csswg.org/css-values-3/#funcdef-calc">calc()</a>. That is, <a href="https://drafts.csswg.org/css2/visudet.html#propdef-width">width: calc(0 + 5px);</a> is invalid, even though both <a href="https://drafts.csswg.org/css2/visudet.html#propdef-width">width: 0;</a> and <a href="https://drafts.csswg.org/css2/visudet.html#propdef-width">width: 5px;</a> are valid.

stylelint isn't validating CSS to match W3C specs. Validating `calc()` will require a lot of algorithms. There is a plugin [stylelint-csstree-validator](https://github.com/csstree/stylelint-validator) which validates declaration values to match W3C specs.

I don't think we should change `length-zero-no-unit`. I would suggest using `stylelint-csstree-validator` to catch invalid values.

> As a workaround, I've converted all of the `calc()` that were subtracting from `0px` to multiply their results by `-1` instead.

Good solution :)

**Comment by Kovensky** [comment]
The problem here is that `length-zero-no-unit` asks you to change your code to invalid syntax to appease it 😢

**Comment by hudochenkov** [comment]
Yeah, but it can't know if this specific situation is valid without heavy algorithms for checking CSS validation.

**Comment by jeddy3** [comment]
> Yeah, but it can't know if this specific situation is valid without heavy algorithms for checking CSS validation.

I agree with that. However, we could make `length-zero-no-unit` more permissive with an option e.g. `ignore: ["inside-calc"]`. Users can then choose whether to use `/* stylelint-disable */` _or_ this new option. At the moment, as @Kovensky said, `length-zero-no-unit` asks users to use invalid syntax to appease it forcing either a workaround in code or by using the disable commands. Adding an option seems like a better third way for the user.

**Comment by jeddy3** [comment]
Labelling up as a new option as it feels like a better third way.

@Kovensky Please consider contributing this option if you'll find it useful. There's [a section in the Developer Guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule) on how to get started.

**Comment by evilebottnawi** [comment]
@jeddy3 i think it is bug when new option, rule doesn't throw warn/error in this case :thinking:

**Comment by jeddy3** [comment]
As @hudochenkov said

> Yeah, but it can't know if this specific situation is valid without heavy algorithms for checking CSS validation.

@evilebottnawi However, if you can fix this _without the heavy lifting_ then that would be ideal. If not, then the `ignore: ["inside-calc"]` option is a lightweight solution.

**Comment by Ephys** [comment]
`ignore: ["inside-calc"]` would be great, as it's becoming an issue when using a fallback value in `var`:

```css
.foo {
  bottom: calc(var(--spacing, 0px) + 12px);
}
```

**Comment by evilebottnawi** [comment]
Not sure it is feature, looks like a bug, because in `calc` you need use `0px` in many case otherwise it is invalid calc, so i think we need ignore `calc` in this rule in favor https://stylelint.io/user-guide/rules/function-calc-no-invalid/

**Comment by jeddy3** [comment]
> so i think we need ignore calc in this rule in favor https://stylelint.io/user-guide/rules/function-calc-no-invalid/

That makes sense to me. There's an overlap between these rules and I think this rule should now ignore `calc` functions outright.
